Add metrics to semaphores

psarna opened the pull request:
This series addresses various issues with metrics and semaphores - it mainly adds missing metrics, which makes it possible to see the length of the queues attached to the semaphores. In case of view building and view update generation, metrics was not present in these services at all, so a first, basic implementation is added.

More precise semaphore metrics would ease the testing and development of load shedding and admission control.

amnonh commented on db/view/view.cc:
```diff
@@ -1213,6 +1213,25 @@ view_builder::view_builder(database& db, db::system_distributed_keyspace& sys_di
+void view_builder::setup_metrics() {
+    namespace sm = seastar::metrics;
+
+    _metrics.add_group("view_builder", {
```
Please use the method form that the function/param is the last parameters

psarna replied:
I would gladly, but `sm::make_queue_length` does not have this form at all, and I'd prefer to use `make_queue_length` insted of the generic `gauge`, because it's more descriptive.

psarna replied:
... and I added the rest in the same form in order not to mix forms in a single call.

nyh commented on table.cc:
```diff
@@ -962,7 +962,9 @@ void table::set_metrics() {
                 ms::make_gauge("live_sstable", ms::description("Live sstable count"), _stats.live_sstable_count)(cf)(ks),
-                ms::make_gauge("pending_compaction", ms::description("Estimated number of compactions pending for this column family"), _stats.pending_compactions)(cf)(ks)
+                ms::make_gauge("pending_compaction", ms::description("Estimated number of compactions pending for this column family"), _stats.pending_compactions)(cf)(ks),
+                ms::make_queue_length("pending_sstable_deletions", [this] { return _sstable_deletion_sem.waiters(); },
```
A thought about naming:
Maybe we should name all the queue length metrics in some common way so they will be easier to recognize as such? I saw you added one metric called "something_queue_length" and here "pending_something", for example.

psarna replied:
I tried to be consistent locally - e.g. when adding a queue length here, I used "pending", but in other metric groups you can also find "queued_something" or "something_current"... and changing existing metrics is out of the question, so there's not that much I can do. I think that the most important thing is to use `ms::make_queue_length` instead of this generic `ms::gauge` - then, when you fetch `localhost:9180/metrics`, this metric is already marked as "queue_length", which is very descriptive.

amnonh commented:
On Tuesday, August 11, 2020, Piotr Sarna <notifications@github.com> wrote:

> *@psarna* commented on this pull request.
> ------------------------------
>
> In table.cc
> <https://github.com/scylladb/scylla/pull/7018#discussion_r468514412>:
>
> > @@ -962,7 +962,9 @@ void table::set_metrics() {
>                  ms::make_gauge("live_disk_space", ms::description("Live disk space used"), _stats.live_disk_space_used)(cf)(ks),
>                  ms::make_gauge("total_disk_space", ms::description("Total disk space used"), _stats.total_disk_space_used)(cf)(ks),
>                  ms::make_gauge("live_sstable", ms::description("Live sstable count"), _stats.live_sstable_count)(cf)(ks),
> -                ms::make_gauge("pending_compaction", ms::description("Estimated number of compactions pending for this column family"), _stats.pending_compactions)(cf)(ks)
> +                ms::make_gauge("pending_compaction", ms::description("Estimated number of compactions pending for this column family"), _stats.pending_compactions)(cf)(ks),
> +                ms::make_queue_length("pending_sstable_deletions", [this] { return _sstable_deletion_sem.waiters(); },
>
> I tried to be consistent locally - e.g. when adding a queue length here, I
> used "pending", but in other metric groups you can also find
> "queued_something" or "something_current"... and changing existing metrics
> is out of the question, so there's not that much I can do. I think that the
> most important thing is to use ms::make_queue_length instead of this
> generic ms::gauge - then, when you fetch localhost:9180/metrics, this
> metric is already marked as "queue_length", which is very descriptive.
>
Not really, the queue length is for collected it remains for backward
compatibility, the name should describe what the metrics id

> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scylladb/scylla/pull/7018#discussion_r468514412>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAQFDP4T7XOTI4MHXP2Z2CDSAEUOJANCNFSM4P25KWJQ>
> .
>

psarna commented:
On 8/11/20 1:58 PM, Amnon Heiman wrote:
> On Tuesday, August 11, 2020, Piotr Sarna <notifications@github.com> wrote:
>
> > *@psarna* commented on this pull request.
> > ------------------------------
> >
> > In table.cc
> > <https://github.com/scylladb/scylla/pull/7018#discussion_r468514412>:
> >
> > > @@ -962,7 +962,9 @@ void table::set_metrics() {
> > ms::make_gauge("live_disk_space", ms::description("Live disk space 
> used"), _stats.live_disk_space_used)(cf)(ks),
> > ms::make_gauge("total_disk_space", ms::description("Total disk space 
> used"), _stats.total_disk_space_used)(cf)(ks),
> > ms::make_gauge("live_sstable", ms::description("Live sstable 
> count"), _stats.live_sstable_count)(cf)(ks),
> > - ms::make_gauge("pending_compaction", ms::description("Estimated 
> number of compactions pending for this column family"), 
> _stats.pending_compactions)(cf)(ks)
> > + ms::make_gauge("pending_compaction", ms::description("Estimated 
> number of compactions pending for this column family"), 
> _stats.pending_compactions)(cf)(ks),
> > + ms::make_queue_length("pending_sstable_deletions", [this] { return 
> _sstable_deletion_sem.waiters(); },
> >
> > I tried to be consistent locally - e.g. when adding a queue length 
> here, I
> > used "pending", but in other metric groups you can also find
> > "queued_something" or "something_current"... and changing existing 
> metrics
> > is out of the question, so there's not that much I can do. I think 
> that the
> > most important thing is to use ms::make_queue_length instead of this
> > generic ms::gauge - then, when you fetch localhost:9180/metrics, this
> > metric is already marked as "queue_length", which is very descriptive.
> >
> Not really, the queue length is for collected it remains for backward
> compatibility, the name should describe what the metrics id
Oh, ok, then I completely misunderstood what this utility function is 
for. I'll just use gauge then, and switch to the "right" order of 
parameters in v2.
>
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/scylladb/scylla/pull/7018#discussion_r468514412>, or
> > unsubscribe
> > 
> <https://github.com/notifications/unsubscribe-auth/AAQFDP4T7XOTI4MHXP2Z2CDSAEUOJANCNFSM4P25KWJQ>
> > .
> >
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/scylladb/scylla/pull/7018#issuecomment-671901900>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ACPTEFZEDHDNP2JU6H5Z3FLSAEW6HANCNFSM4P25KWJQ>.
>

nyh commented:
> Oh, ok, then I completely misunderstood what this utility function is for. I'll just use gauge then, and switch to the "right" order of parameters in v2.

@amnonh @psarna can you please add some comments or deprecation tags on specific functions which you don't want anyone to use, and explain why?
Also @psarna made a good point that this "queue_length" thing is written in the metric as a comment, which allows a user to get some more insights on the different types of metrics. Why should we not have this feature, unrelated to collected but simply as a still-useful feature that is not deprecated?

psarna commented:
v2:
 - dropped `make_queue_length` calls since I heard they're deprecated and reordered the parameters to the "right" way
 - changed the names to make all metrics which represent queue lengths start with `pending_`

psarna commented:
@nyh I personally don't really care which order of parameters is used, as long as the series gets merged and I can use the metrics later :) that was my sole reason for sending v2. In v1, I used `make_queue_length` because I understand what it means the moment I look at it, which is not the case with `make_gauge` (and I only recently discovered how to pronounce "gauge" correctly). If we decide that `make_queue_length` is, in fact, **not** deprecated, I'd be happy to bring it back, but if the choice is "gauge" vs a long fight over naming, I choose gauge.

nyh commented:
> @nyh I personally don't really care which order of parameters is used, as long as the series gets merged and I can use the metrics later :) that was my sole reason for sending v2. In v1, I used `make_queue_length` because I understand what it means the moment I look at it, which is not the case with `make_gauge` (and I only recently discovered how to pronounce "gauge" correctly). If we decide that `make_queue_length` is, in fact, **not** deprecated, I'd be happy to bring it back, but if the choice is "gauge" vs a long fight over naming, I choose gauge.

I'll review your new version. @amnonh please do add deprecation warnings and/or comments in front of the functions that you consider deprecated and don't like to see used in patches - so that the next guy who comes along to add metrics doesn't make the same "mistakes" as Piotr did here. 
By the way, last time I added metrics (for Alternator), I also wrote the document `docs/metrics.md`. Piotr, you can find there my understanding of what "gauge" means for us, and see there is indeed no mention of a "queue length" type of metric.  I also wanted to add a section on how to add new metrics in the code, but never got around to doing it, unfortunately.

nyh commented:
Thanks. Merged to next: 813564790660f236e2efcaacfa5f26120bad050d
